### PR TITLE
Oxidize available device

### LINKED ISSUE #3864: Oxidize `AvailableDevice`, `get_available_device`, `get_key_file`, `list_available_devices`, `load_device_file`

### PR DESCRIPTION
**Pull request by Volham22**
# What has changed ?

This PR closes #3864 

- [ ] Does this PR affect the User ?

### CODE REVIEW

**Comment by Volham22** [comment]
We need #3887 to be merged before. Some function here depends on function oxidized in this PR.

File: oxidation/libparsec/crates/client_types/src/local_device_file.rs [hunk 1]
```diff
@@ -94,12 +96,44 @@ impl DeviceFile {
...
 
-        rmp_serde::from_slice::<DeviceFile>(&data)
-            .map_err(|_| LocalDeviceError::Deserialization(key_file_path.to_path_buf()))
+        let result = rmp_serde::from_slice::<DeviceFile>(&data)
+            .map_err(|_| LocalDeviceError::Deserialization(key_file_path.to_path_buf()));
+
+        // In case of failure try to load a legacy_device and convert it to a
+        // regular device file
+        if result.is_err() {
+            Self::load_legacy_device_file(key_file_path, &data)
```
Comment (TimeEngineer) [inline oxidation/libparsec/crates/client_types/src/local_device_file.rs:109]: You can use https://doc.rust-lang.org/std/result/enum.Result.html#method.or_else

File: oxidation/libparsec/crates/client_types/src/local_device_file.rs [hunk 2]
```diff
@@ -275,7 +336,10 @@ pub fn list_available_devices(config_dir: &Path) -> LocalDeviceResult<Vec<Availa
             // Load the device file
             Ok(device) => device,
             // Ignore invalid files
-            Err(_) => continue,
+            Err(e) => {
+                println!("Invalid {:?}", e);
```
Comment (TimeEngineer) [inline oxidation/libparsec/crates/client_types/src/local_device_file.rs:340]: Remove `println!`

**Review by TimeEngineer — CHANGES_REQUESTED** [review]

File: oxidation/libparsec/crates/client_types/src/local_device_file.rs (hunk 1 above)
Comment (FirelightFlagboy) [reply oxidation/libparsec/crates/client_types/src/local_device_file.rs:109]: Strange `clippy` don't say anything about that :thinking:

Comment (Volham22) [reply oxidation/libparsec/crates/client_types/src/local_device_file.rs:109]: Done.

File: oxidation/libparsec/crates/client_types/src/local_device_file.rs (hunk 2 above)
Comment (Volham22) [reply oxidation/libparsec/crates/client_types/src/local_device_file.rs]: Done.

File: oxidation/libparsec/crates/client_types/src/local_device_file.rs [hunk 3]
```diff
@@ -94,12 +96,40 @@ impl DeviceFile {
...
             .map_err(|_| LocalDeviceError::Access(key_file_path.to_path_buf()))?;
 
-        rmp_serde::from_slice::<DeviceFile>(&data)
-            .map_err(|_| LocalDeviceError::Deserialization(key_file_path.to_path_buf()))
+        let result = rmp_serde::from_slice::<DeviceFile>(&data)
+            .map_err(|_| LocalDeviceError::Deserialization(key_file_path.to_path_buf()));
+
+        // In case of failure try to load a legacy_device and convert it to a
+        // regular device file
+        result.or_else(|_| Self::load_legacy_device_file(key_file_path, &data))
```
Comment (TimeEngineer) [inline oxidation/libparsec/crates/client_types/src/local_device_file.rs:108]: You can chain it

**Review by TimeEngineer — APPROVED** [review]

File: oxidation/libparsec/crates/client_types/src/local_device.rs [hunk 4]
```diff
@@ -49,12 +49,7 @@ impl LocalDevice {
         key_file: &Path,
         password: &str,
     ) -> Result<Self, LocalDeviceError> {
-        let ciphertext = std::fs::read(key_file)
-            .map_err(|_| LocalDeviceError::Access(key_file.to_path_buf()))?;
-        let device_file: DeviceFile = match rmp_serde::from_slice(&ciphertext) {
-            Ok(result) => result,
-            Err(_) => todo!("Handle legacy device"),
-        };
+        let device_file = DeviceFile::load(key_file)?;
```
[the remark this answers is not in the document]
Comment (Volham22) [reply oxidation/libparsec/crates/client_types/src/local_device.rs:52]: Is the new `load` function merged?

[the remark this answers is not in the document]
Comment (TimeEngineer) [reply oxidation/libparsec/crates/client_types/src/local_device.rs:52]: Nope, but this is only a remark

**Pull request merged**